cylc/flow/ws_data_mgr.py

```diff
@@ -776,17 +747,23 @@ def update_workflow_statuses(self):
         workflow.stamp = f'{self.workflow_id}@{update_time}'
         workflow.last_updated = update_time
 
-        state_count_totals = {}
-        for _, count in list(self.state_count_cycles.items()):
-            for state, state_count in count.items():
-                state_count_totals.setdefault(state, 0)
-                state_count_totals[state] += state_count
-        for state, count in state_count_totals.items():
-            setattr(workflow.state_totals,
-                    copy(state).replace('-', '_'), count)
+        all_states = [
+            t.state
+            for t in self.task_proxies.values()
+            if t.state != ''
+        ]
+
+        workflow.states[:] = set(all_states)
```

Comment:
Can get the list of states from the `Counter` object to avoid iterating over the full list of states twice (as the number of tasks can get quite large).

Reply:
Oh yeah! True  XD

Reply:
Fixed.. (I hope that's what you meant!)